pkg/dynamicinstrumentation/ebpf/ebpf.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2016-present Datadog, Inc.
+
+// Package ebpf provides utility for setting up and instrumenting the bpf code
+// used by dynamic instrumentation
+package ebpf
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"text/template"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/diagnostics"
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/bytecode/runtime"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var (
+	bpffs = "/sys/fs/bpf" //TODO: Check via `mount(2)`
+)
+
+var (
+	globalTempDirPath string
+	globalHeadersPath string
+)
+
+// SetupEventsMap creates the ringbuffer which all programs will use for sending output
+func SetupEventsMap() error {
+	var err error
+	events, err := ebpf.NewMap(&ebpf.MapSpec{
+		Name:       "events",
+		Type:       ebpf.RingBuf,
+		MaxEntries: 1 << 24,
+	})
+	if err != nil {
+		return fmt.Errorf("could not create bpf map for sharing events with userspace: %w", err)
+	}
+	ditypes.EventsRingbuffer = events
+	return nil
+}
+
+// SetupHeaders sets up the needed header files for probes in a temporary directory
+func SetupHeaders() error {
+	var err error
+	globalTempDirPath, err = os.MkdirTemp("/tmp", "dd-go-di*")
+	if err != nil {
+		return err
+	}
+
+	tempDir, err := os.MkdirTemp(globalTempDirPath, "run-*")
+	if err != nil {
+		return fmt.Errorf("couldn't create temp directory: %w", err)
+	}
+
+	headersPath, err := loadHeadersToTmpfs(tempDir)
+	if err != nil {
+		return fmt.Errorf("couldn't load headers to tmpfs: %w", err)
+	}
+
+	globalHeadersPath = headersPath
+	return nil
+}
+
+// AttachBPFUprobe attaches the probe to the specified process
+func AttachBPFUprobe(procInfo *ditypes.ProcessInfo, probe *ditypes.Probe) error {
+	executable, err := link.OpenExecutable(procInfo.BinaryPath)
+	if err != nil {
+		diagnostics.Diagnostics.SetError(procInfo.ServiceName, procInfo.RuntimeID, probe.ID, "ATTACH_ERROR", err.Error())
+		return fmt.Errorf("could not open proc executable for attaching bpf probe: %w", err)
+	}
+
+	spec, err := ebpf.LoadCollectionSpecFromReader(probe.InstrumentationInfo.BPFObjectFileReader)
+	if err != nil {
+		diagnostics.Diagnostics.SetError(procInfo.ServiceName, procInfo.RuntimeID, probe.ID, "ATTACH_ERROR", err.Error())
+		return fmt.Errorf("could not create bpf collection for probe %s: %w", probe.ID, err)
+	}
+
+	mapReplacements := map[string]*ebpf.Map{}
+	if probe.ID != ditypes.ConfigBPFProbeID {
+		// config probe is special and should not be on the same ringbuffer
+		// as the rest of regular events. Despite having the same "events" name,
+		// not using the pinned map means the config program uses a different
+		// ringbuffer.
+		mapReplacements["events"] = ditypes.EventsRingbuffer
+	} else {
+		configEvents, err := ebpf.NewMap(&ebpf.MapSpec{
```

Comment:
This is creating a ring buffer for each uprobe. Did you want to share one ringbuf between all uprobes?